go/engine/puk_upkeep.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+// PerUserKeyUpkeep rolls the user's per-user-key if the last PUK
+// was added by a now-revoked device.
+// Does not add a first per-user-key. Use PerUserKeyUpgrade for that.
+// This engine makes up for the fact that after a self-deprovision
+// the latest PUK for a user was generated on the very machine they
+// wanted to deprovision.
+// This will not notice if a device revoked another device but neglected
+// to roll the PUK. No clients should do that.
+package engine
+
+import (
+	"errors"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+// PerUserKeyUpkeep is an engine.
+type PerUserKeyUpkeep struct {
+	libkb.Contextified
+	args       *PerUserKeyUpkeepArgs
+	DidRollKey bool
+}
+
+type PerUserKeyUpkeepArgs struct{}
+
+// NewPerUserKeyUpkeep creates a PerUserKeyUpkeep engine.
+func NewPerUserKeyUpkeep(g *libkb.GlobalContext, args *PerUserKeyUpkeepArgs) *PerUserKeyUpkeep {
+	return &PerUserKeyUpkeep{
+		args:         args,
+		Contextified: libkb.NewContextified(g),
+	}
+}
+
+// Name is the unique engine name.
+func (e *PerUserKeyUpkeep) Name() string {
+	return "PerUserKeyUpkeep"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *PerUserKeyUpkeep) Prereqs() Prereqs {
+	return Prereqs{
+		Session: true,
+	}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *PerUserKeyUpkeep) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *PerUserKeyUpkeep) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+// Run starts the engine.
+func (e *PerUserKeyUpkeep) Run(ctx *Context) (err error) {
+	defer e.G().CTrace(ctx.GetNetContext(), "PerUserKeyUpkeep", func() error { return err })()
+	return e.inner(ctx)
+}
+
+func (e *PerUserKeyUpkeep) inner(ctx *Context) error {
+	e.G().Log.CDebugf(ctx.GetNetContext(), "PerUserKeyUpkeep load self")
+
+	uid := e.G().GetMyUID()
+	if uid.IsNil() {
+		return libkb.NoUIDError{}
+	}
+
+	loadArg := libkb.NewLoadUserArgBase(e.G()).
+		WithNetContext(ctx.GetNetContext()).
+		WithUID(uid).
+		WithSelf(true).
+		WithPublicKeyOptional()
+	upak, me, err := e.G().GetUPAKLoader().LoadV2(*loadArg)
+	if err != nil {
+		return err
+	}
+	// `me` could be nil.
```

Comment:
That's what upak loader does. It returns the upak for sure but the user only if it happened to load one. I got burned by this once so that's how I remember.